### PR TITLE
gitlab_project - Add ability to create project under a user

### PR DESCRIPTION
**Pull request by stgrace**
##### SUMMARY

We wanted to create gitlab projects per user, for a use case we have. However, it is only possible to create projects under a group or under the owner of the gitlab api token.

This pull request adds the functionality to create a project under different users using their `username`.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

gitlab_project
##### ADDITIONAL INFORMATION

```paste below
TASK [Create project] ****************************************************************************************************************
changed: [localhost] => {"changed": true, "msg": "Successfully created or updated the project my_first_project2", "project": {"_links": {"events": "https://git.dome.dev/api/v4/projects/544/events", "labels": "https://git.dome.dev/api/v4/projects/544/labels", "members": "https://git.dome.dev/api/v4/projects/544/members", "merge_requests": "https://git.dome.dev/api/v4/projects/544/merge_requests", "repo_branches": "https://git.dome.dev/api/v4/projects/544/repository/branches", "self": "https://git.dome.dev/api/v4/projects/544"}, "allow_merge_on_skipped_pipeline": null, "archived": false, "auto_cancel_pending_pipelines": "enabled", "auto_devops_deploy_strategy": "continuous", "auto_devops_enabled": false, "autoclose_referenced_issues": true, "avatar_url": null, "build_coverage_regex": null, "build_git_strategy": "fetch", "build_timeout": 3600, "builds_access_level": "enabled", "can_create_merge_request_in": true, "ci_config_path": "", "ci_default_git_depth": 50, "container_expiration_policy": {"cadence": "1d", "enabled": true, "keep_n": 10, "name_regex": null, "name_regex_keep": null, "next_run_at": "2021-06-18T11:29:31.376+02:00", "older_than": "90d"}, "container_registry_enabled": true, "created_at": "2021-06-17T11:29:31.352+02:00", "creator_id": 2, "default_branch": "main", "description": null, "emails_disabled": null, "empty_repo": true, "forking_access_level": "enabled", "forks_count": 0, "http_url_to_repo": "https://git.dome.dev/wart.spitaels/my_first_project2.git", "id": 544, "import_error": null, "import_status": "none", "issues_access_level": "disabled", "issues_enabled": false, "jobs_enabled": true, "last_activity_at": "2021-06-17T11:29:31.352+02:00", "lfs_enabled": false, "merge_method": "rebase_merge", "merge_requests_access_level": "enabled", "merge_requests_enabled": true, "name": "my_first_project2", "name_with_namespace": "wart.spitaels / my_first_project2", "namespace": {"avatar_url": "https://secure.gravatar.com/avatar/a25512d88124194ff3bed9e4d2e08586?s=80&d=identicon", "full_path": "wart.spitaels", "id": 10, "kind": "user", "name": "wart.spitaels", "parent_id": null, "path": "wart.spitaels", "web_url": "https://git.dome.dev/wart.spitaels"}, "only_allow_merge_if_all_discussions_are_resolved": false, "only_allow_merge_if_pipeline_succeeds": false, "owner": {"avatar_url": "https://secure.gravatar.com/avatar/a25512d88124194ff3bed9e4d2e08586?s=80&d=identicon", "id": 5, "name": "wart.spitaels", "state": "active", "username": "wart.spitaels", "web_url": "https://git.dome.dev/wart.spitaels"}, "packages_enabled": true, "pages_access_level": "private", "path": "my_first_project2", "path_with_namespace": "wart.spitaels/my_first_project2", "printing_merge_request_link_enabled": true, "public_jobs": true, "readme_url": null, "remove_source_branch_after_merge": true, "repository_access_level": "enabled", "repository_storage": "default", "request_access_enabled": true, "resolve_outdated_diff_discussions": false, "runners_token": "zGTWfD7LQi5fb9y-Ltoe", "service_desk_address": null, "service_desk_enabled": false, "shared_runners_enabled": true, "shared_with_groups": [], "snippets_access_level": "enabled", "snippets_enabled": true, "ssh_url_to_repo": "git@git.dome.dev:wart.spitaels/my_first_project2.git", "star_count": 0, "suggestion_commit_message": null, "tag_list": [], "visibility": "private", "web_url": "https://git.dome.dev/wart.spitaels/my_first_project2", "wiki_access_level": "enabled", "wiki_enabled": true}}
```

### CODE REVIEW

**Comment by russoz** [comment]
Thanks for the contribution @stgrace !

File: changelogs/fragments/2824-gitlab_project-project-under-user.yml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+---
+minor_changes:
+  - User projects are created using namespace ID, instead of user ID.
```
Comment (russoz) [inline changelogs/fragments/2824-gitlab_project-project-under-user.yml:3]: Please adjust the changelog fragment:
```suggestion
  - gitlab_project - user projects are created using namespace ID, instead of user ID (https://github.com/ansible-collections/community.general/pull/2824).
```
Same for the other line.

Comment (stgrace) [reply changelogs/fragments/2824-gitlab_project-project-under-user.yml]: Should be fixed in the latest commit

File: changelogs/fragments/2824-gitlab_project-project-under-user.yml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+---
+minor_changes:
+  - gitlab_project - user projects are created using namespace ID, instead of user ID (https://github.com/ansible-collections/community.general/pull/2824).
+  - gitlab_project - projects can be created under other user's namespaces (https://github.com/ansible-collections/community.general/pull/2824).
```
Comment (felixfontein) [inline changelogs/fragments/2824-gitlab_project-project-under-user.yml:4]: I'm not sure why this change needs two separate changelog fragments. What is the relation between these fragments to each other, and to this change (adding a new module option, which isn't mentioned in any of them)?

Comment (felixfontein) [reply changelogs/fragments/2824-gitlab_project-project-under-user.yml:4]: @stgrace ^

Comment (stgrace) [reply changelogs/fragments/2824-gitlab_project-project-under-user.yml:4]: So first thing we did was try out the module. My colleague was trying the module out with a personal access key made under his account, and somehow projects were being created under another colleagues account. Turns out the user ID for my colleague was 11, and his workspace ID was 16. The user ID of the other colleague under which the projects were being created was 6, but her workspace ID was 11. 

TL;DR
The module was picking the user ID, but should have been picking the user's workspace ID. We fixed this, and added an option to create projects under other user's workspaces.

File: plugins/modules/source_control/gitlab/gitlab_project.py [hunk 3]
```diff
@@ -109,6 +109,12 @@
     required: false
     default: false
     version_added: "2.0.0"
+  username:
+    description:
+      - Used to create a personal project under a user's name.
+    type: str
+    required: false
```
Comment (aminvakil) [inline plugins/modules/source_control/gitlab/gitlab_project.py:116]:
```suggestion
```

**Review by aminvakil — COMMENTED** [review]
Thanks for adding this!

File: changelogs/fragments/2824-gitlab_project-project-under-user.yml (hunk 2 above)
Comment (felixfontein) [reply changelogs/fragments/2824-gitlab_project-project-under-user.yml:4]: In that case, one is a bugfix, while the other is a feature. These should be two separate PRs (because one can be backported to stable-2 and the other not), and the bugfix changelog fragment should be in the `bugfixes` section.

Anyway, I don't see how this actually changes something in the case `username` is not provided. `namespace` is set in line 374 the same way as before.

Comment (stgrace) [reply changelogs/fragments/2824-gitlab_project-project-under-user.yml:4]: The getting of the namespace is done with a different ID that is set in the changes we made. The variable name did not change but the value did. Before it was retrieved using UID, now it is using the user's workspace ID.

I can make different PR's for it

Comment (felixfontein) [reply changelogs/fragments/2824-gitlab_project-project-under-user.yml:4]: Ah, I missed the `gitlab_instance.users.list` → `gitlab_instance.namespaces.list`. In any case, splitting this up in two PRs is a good idea (I'd create a new PR for the bugfix, and keep this one for the feature as a WIP until the bugfix PR is merged).

But shouldn't `user_group_id` be renamed to `namespace_id` then? And why not just keep the namespace, instead of retrieving it a second time shortly after?

Also I'm a bit worried on two things:
1. You are listing workspaces when searching for the user name, instead of listing users. Is that really the correct approach?
2. What happens to the `group_identifier` branch? There, `user_group_id` is set to `group.id`. Is this also a namespace ID?

Comment (stgrace) [reply changelogs/fragments/2824-gitlab_project-project-under-user.yml:4]: I'll make a new PR and continue the discussion there

**Comment by felixfontein** [comment]
Will need a rebase once #2881 is merged.

**Comment by felixfontein** [comment]
This now needs a rebase.

**Comment by stgrace** [comment]
@felixfontein I'll get to it

File: changelogs/fragments/2824-gitlab_project-project-under-user.yml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+---
+minor_changes:
+  - User projects are created using namespace ID, instead of user ID.
+  - Projects can be created under other user's namespaces.
```
Comment (felixfontein) [inline changelogs/fragments/2824-gitlab_project-project-under-user.yml:4]:
```suggestion
  - gitlab_project - projects can be created under other user's namespaces with the new ``username`` option (https://github.com/ansible-collections/community.general/pull/2824).
```

**Review by felixfontein — COMMENTED** [review]
I think it's a good idea to add `['group', 'username']` to `mutually_exclusive`, since they cannot be used together.

**Comment by stgrace** [comment]
> I think it's a good idea to add `['group', 'username']` to `mutually_exclusive`, since they cannot be used together.

Good idea, done

**Review by aminvakil — APPROVED** [review]
LGTM

**Review by felixfontein — APPROVED** [review]

**Comment by aminvakil** [comment]
@stgrace Thanks for adding this feature!
@russoz @felixfontein Thanks for reviewing!
